### ISSUE TITLE
ResourceExhaustedError during retraining in Object detection API

### ISSUE THREAD

**Issue by Garima-Mishra** [tensorflow/models#1766]
I am trying to train the mobilenet v1 with pets data set.
from tensorflow.python.client import device_lib
device_lib.list_local_devices()
gives following output 
name: GeForce GTX 750 Ti
major: 5 minor: 0 memoryClockRate (GHz) 1.15
pciBusID 0000:03:00.0
Total memory: 1.95GiB
Free memory: 1.93GiB

But when I run training. I get the error 
tensorflow/core/framework/op_kernel.cc:1158] Resource exhausted: OOM when allocating tensor with shape[24,128,75,75]
INFO:tensorflow:Error reported to Coordinator: <class 'tensorflow.python.framework.errors_impl.ResourceExhaustedError'>, OOM when allocating tensor with shape[24,128,75,75]

Even tried with pascal dataset and other pre-trained models also. I was able to run call models for running the object detection tutorial.
I followed the instruction given for running_locally.md
One thing I observed that all of the GPU memory get exhausted before throwing the error.
What is the minimum GPU memory required for training?

**Comment by jch1** [comment]
Hi @Garima-Mishra - most of out models require much more than 2Gb of memory (usually between >10 Gb at training time).  A few things you could do (besides get a new GPU):
1) decrease the training resolution
2) decrease the batch size

In each of these cases you may need to experiment a bit with hyperparameters like learning rate.

**Comment by Garima-Mishra** [comment]
Hi, @jch1 
Thanks for the quick response. Can I use multiple GPU's for training as I have two 8GB GPU as well?
I will try with your suggestions of reducing batch size and resolution as well.

**Comment by jch1** [comment]
Yes, you have to tweak the num_clones parameter (see the object_detection/train.py file).

**Comment by SimsGautam** [comment]
simply tweaking the num_clones parameter (along with adjusting batch size) doesn't seem to get multi-GPU to work @jch1, see #1782

**Comment by tmattio** [comment]
I have the same issue running Faster-RCNN with Inception Resnet V2.
I have been running the model on a GeForce GTX 1080 without any problem, but running on a Tesla K80 (the GPU on the AWS P2 instances) gives me a ResourceExhaustedError.

I've tried to change the resolution, down to 400:

```
image_resizer {
  keep_aspect_ratio_resizer {
    min_dimension: 400
    max_dimension: 400
  }
}
```

And change the batch_size, down to 8

```
second_stage_batch_size: 16
```

But I still have the same error.

**Comment by tmattio** [comment]
Actually, it's not exactly the same issue. Here is the error:

```
ResourceExhaustedError (see above for traceback): Ran out of GPU memory when allocating 0 bytes for 
         [[Node: Loss/RPNLoss/Loss_1/SoftmaxCrossEntropyWithLogits = SoftmaxCrossEntropyWithLogits[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/gpu:0"](Loss/RPNLoss/Loss_1/Reshape_2, Loss/RPNLoss/Loss_1/Reshape_3)]]
         [[Node: gradients/FirstStageFeatureExtractor/InceptionResnetV2/InceptionResnetV2/Repeat_1/block17_11/Conv2d_1x1/convolution_grad/tuple/control_dependency_1/_5227 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_28924_gradients/FirstStageFeatureExtractor/InceptionResnetV2/InceptionResnetV2/Repeat_1/block17_11/Conv2d_1x1/convolution_grad/tuple/control_dependency_1", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
```

The `Ran out of GPU memory when allocating 0 bytes` is interesting...

Any ideas?

**Comment by sbmaruf** [comment]
I am new to tensorflow and Machine Learning. Recently I am working on a model. My model is like below,

1. Character level Embedding Vector -> Embedding lookup -> LSTM1

2. Word level Embedding Vector->Embedding lookup -> LSTM2 

3. [LSTM1+LSTM2] -> single layer MLP-> softmax layer

4. [LSTM1+LSTM2] -> Single layer MLP-> WGAN discriminator

while I'm working on this model I got the following error. I thought My batch is too big. Thus I tried to reduce the batch size from 20 to 10 but it doesn't work. 

> ResourceExhaustedError (see above for traceback): OOM when allocating
> tensor with shape[24760,100] 	 [[Node:
> chars/bidirectional_rnn/bw/bw/while/bw/lstm_cell/split =
> Split[T=DT_FLOAT, num_split=4,
> _device="/job:localhost/replica:0/task:0/device:GPU:0"](gradients_2/Add_3/y,
> chars/bidirectional_rnn/bw/bw/while/bw/lstm_cell/BiasAdd)]] 	 [[Node:
> bi-lstm/bidirectional_rnn/bw/bw/stack/_167 =
> _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0",
> send_device="/job:localhost/replica:0/task:0/device:GPU:0",
> send_device_incarnation=1,
> tensor_name="edge_636_bi-lstm/bidirectional_rnn/bw/bw/stack",
> tensor_type=DT_INT32,
> _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

tensor with ***shape[24760,100]*** means 2476000*32/1024*1024 = 75.*** MB memory. I am running the code on a titan X(11 GB) gpu. What could go wrong. Why this type of error occured?

Extra info: the size of the LSTM1 is 100. for bidirectional LSTM it becomes 200.
The size of the LSTM2 is 300. For Bidirectional LSTM it becomes 600.

**Comment by sbmaruf** [comment]
https://stackoverflow.com/questions/48007984/training-broke-with-resourceexausted-error


follow the answer.

**Comment by anonym24** [comment]
I have NVIDIA GeForce GTX 1060 6 GB and this error happens to me

It's not about GPU memory

I was training on Notebook with NVIDIA GeForce 940 MX 2 GB and it was working ok

**Comment by JigyasaK** [comment]
I had the same issue even when my batch_size was already 1. I resized the image to 500x500 and then it ran without errors.

**Comment by xunnv** [comment]
I will give you a link of issue that is similar to yours: #1817.
For guys who stuck into the same problem with a not-bad GPU, there are some ways that would handle it:
1. reduce your batch_size in train_config block, as mentioned by https://github.com/tensorflow/models/issues/1766#issuecomment-311250580
2. resize your images to a lower resolution , by doing the way like https://github.com/tensorflow/models/issues/1817#issuecomment-325851293
3. add the following lines into object_detection/model_main.py
```python
session_config = tf.ConfigProto()
session_config.gpu_options.per_process_gpu_memory_fraction = 0.5    # replace this field with whichever real number you prefer
                                                                    # also gives a workaround to specify RAM usage
config = tf.estimator.RunConfig(model_dir=FLAGS.model_dir, session_config=session_config)
```
4. try the following snippet as mentioned by https://github.com/tensorflow/models/issues/1817#issuecomment-325988741
```
train_config: {
  batch_size: 1
  ...
  batch_queue_capacity: 150
  num_batch_queue_threads: 8
  prefetch_queue_capacity: 10
}
```
Hopefully at least one of these solutions would help

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
